### PR TITLE
CDAP-1853: Use resource assignment to divide kafka partitions across Metrics Processors

### PR DESCRIPTION
**Pull request by abaranau**

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java [hunk 1]
```diff
@@ -40,21 +58,52 @@
...
+          coordinator = new ResourceCoordinator(zkClient,
+                                                discoveryServiceClient,
+                                                new BalancedAssignmentStrategy());
+          coordinator.start();
+        }
+
+        @Override
+        public void follower() {
+          if (coordinator != null) {
+            coordinator.stop();
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java]: Do you need to wait for stop to complete (meaning stopAndWait)?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java [hunk 2]
```diff
@@ -77,7 +126,22 @@ public void execute(Runnable command) {
...
     LOG.info("Starting Metrics Processor ...");
-    multiElection.start();
+
+    Discoverable discoverable = createDiscoverable(Constants.Service.METRICS_PROCESSOR);
+    cancelDiscoverable = discoveryService.register(ResolvingDiscoverable.of(discoverable));
+
+    election.start();
+    resourceClient.startAndWait();
+
+    int partitionSize = conf.getInt(Constants.Metrics.KAFKA_PARTITION_SIZE,
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java]: So currently the only way to change partition is by changing config, right?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java [hunk 3]
```diff
@@ -93,38 +157,93 @@ protected void triggerShutdown() {
   @Override
   protected void shutDown() throws Exception {
     LOG.info("Stopping Metrics Processor ...");
+    Throwable throwable = null;
+    try {
+      Services.chainStop(election, resourceClient);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java]: call `.get()` to block.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java [hunk 4]
```diff
@@ -93,38 +157,93 @@ protected void triggerShutdown() {
...
+        return serviceName;
+      }
+
+      @Override
+      public InetSocketAddress getSocketAddress() {
+        return finalAddress;
       }
     };
   }
+
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java]: Remove unnecessary new line.

**Comment by chtyim** [comment]
Just couple comments. Rest LGTM.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java (hunk 1 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java]: can I do it in follower()? I mean is it fine to do a blocking call in there?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java (hunk 2 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java]: yes

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java [hunk 5]
```diff
@@ -40,21 +58,52 @@
...
+    this.election =
+      new LeaderElection(zkClient, Constants.Service.METRICS_PROCESSOR, new ElectionHandler() {
+        private ResourceCoordinator coordinator;
+
+        @Override
+        public void leader() {
+          coordinator = new ResourceCoordinator(zkClient,
+                                                discoveryServiceClient,
+                                                new BalancedAssignmentStrategy());
+          coordinator.start();
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java]: should this be coordinator.startAndWait ?

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/runtime/KafkaMetricsProcessorService.java]: yes

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT2215-3

**Comment by abaranau** [comment]
will merge after build re-run

**Pull request merged**